File: libs/language/src/lib/utils/limits/Limits.ts
import { BlueNode } from '../../model';

/**
 * Abstract class for specifying limits on node extension
 */
export abstract class Limits {
  /**
   * Determines if a path segment should be extended
   * @param pathSegment - The path segment
   * @param currentNode - The current node
   * @returns True if the segment should be extended, false otherwise
   */
  public abstract shouldExtendPathSegment(
    pathSegment: string,
    currentNode: BlueNode
  ): boolean;

  /**
   * Determines if a path segment should be merged
   * @param pathSegment - The path segment
   * @param currentNode - The current node
   * @returns True if the segment should be merged, false otherwise
   */
  public abstract shouldMergePathSegment(
    pathSegment: string,
    currentNode: BlueNode
  ): boolean;

  /**
   * Enters a path segment with an optional node
   * @param pathSegment - The path segment
   * @param currentNode - The current node (optional)
   */
  public abstract enterPathSegment(
    pathSegment: string,
    currentNode?: BlueNode
  ): void;

  /**
   * Exits a path segment
   */
  public abstract exitPathSegment(): void;
}

File: references/java-sdk/src/main/java/blue/language/utils/limits/Limits.java
package blue.language.utils.limits;

import blue.language.model.Node;

public interface Limits {

    Limits NO_LIMITS = new NoLimits();

    boolean shouldExtendPathSegment(String pathSegment, Node currentNode);

    boolean shouldMergePathSegment(String pathSegment, Node currentNode);

    default void enterPathSegment(String pathSegment) {
        enterPathSegment(pathSegment, null);
    }

    void enterPathSegment(String pathSegment, Node currentNode);
    void exitPathSegment();
}